Task-49104 Implement Metadata API and Activity Favorites as first example

boubaker opened the pull request.

boubaker committed:
Task-49104 Implement Metadata API Basic operations

boubaker committed:
Task-49104 Enrich Metadata API with additional Basic operations

boubaker committed:
TASK-48997 Add Favorites Feature Flag

boubaker committed:
TASK-49104 Delete Metadata Items when the activity is Deleted

boubaker committed:
TASK-48998 Add Favorite Button on Activity

boubaker committed:
TASK-49012 Add v-alert message when adding/removing activity favorite

boubaker committed:
TASK-49104 Allow to override or ignore existing Metadata Item

boubaker committed:
TASK-49000 List Activity Favorites from Search UI

boubaker committed:
TASK-49001 Display Activity Favorite Action in Search Card

boubaker committed:
Task-49104 Ensure to retrieve distinct values for object Ids

boubaker committed:
TASK-48998 Make Activity Header buttons extensible via extensionRegistry

boubaker committed:
TASK-48998 Define Service and Rest layers on top pf Metadata API

plamarque commented on component/service/src/main/java/org/exoplatform/social/rest/impl/metadata/FavoriteRest.java:
```diff
@@ -0,0 +1,165 @@
+
+import io.swagger.annotations.*;
+
+@Path(VersionResources.VERSION_ONE + "/social/favorites")
```
couldn't we have /favorites without prefix? it's not related to social fesures amd supposed to be transversal even for obbjects outside social

boubaker replied:
Unfortunately we have already a `favorites` endpoint : https://github.com/exoplatform/ecms/blob/develop/core/connector/src/main/java/org/exoplatform/wcm/connector/collaboration/FavoriteRESTService.java#L37

Besides, adding a parent category for REST endpoints ( like 'social' here to reference the module ), will give us more flexibility for routing rules, such as in Microservices Architecture (where we will be able to route all 'social' endpoints to the Monolith, a dedicated Microservice, applying a specific rule ...)

boubaker committed:
TASK-49001 Dispatch Favorite modification from Search card to impact button on Stream

boubaker committed:
TASK-49104 Rename Processor Class and add some comments

boubaker committed:
TASK-49001 Avoid displaying User popover when hovering Favorites button

boubaker committed:
TASK-49001 Allow to attach parentId to the favorite of an object

plamarque reviewed (commented):
i'd prefer /favorites instead of /social/favorites to make it clear that it's a transversal service (not tied to "social" features)

boubaker committed:
TASK-49000 Optimize retrieved activities ids while searching
Prior to this change, the used 'limit' was all time equal to 'SEARCH_FAVORITES_HARD_LIMIT' while when no terms to search, we know exactly how much activities to retrieve.

boubaker committed:
TASK-49000 Include Metadata in Activity indexed document
Prior to this change, the search on favorite activities was made using a Database query first to retrieve the list of user activities made as favorite, then a second query is made to ES to retrieve the activities that matches the user's search term that are favorite as well. The new strategy consists of retrieving the favorites from DB only when the user didn't entered a search term, that way we do a simple listing when the user search isn't about a full Text search. When the user searches the favorites using a search term, a single ES query will be used to combine search about the term & favorites at the same time.

boubaker committed:
TASK-49000 Improve Activity Search Card UI
Prior to this change, the UI of search activity card computes the text to display in Seach card by JS. This modification introduces a new way, using CSS, to display the ellipsis on multiline content.
